cache/sampleStore.js

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * ./cache/sampleStore.js
+ *
+ * Redis Cache
+ */
+'use strict'; // eslint-disable-line strict
+const redisClient = require('./redisCache').client.sampleStore;
+const Sample = require('../db').Sample;
+const featureToggles = require('feature-toggles');
+const PFX = 'samsto';
+const SEP = ':';
+const constants = {
+  featureName: 'enableRedisSampleStore',
+  fieldsToStringify: {
+    aspect: [
+      'relatedLinks',
+      'tags',
+      'criticalRange',
+      'warningRange',
+      'infoRange',
+      'okRange',
+    ],
+    sample: ['relatedLinks'],
+  },
+  indexKey: {
+    aspect: PFX + SEP + 'aspects',
+    sample: PFX + SEP + 'samples',
+    subject: PFX + SEP + 'subjects',
+  },
+  objectType: { aspect: 'aspect', sample: 'sample', subject: 'subject' },
+  prefix: PFX,
+  separator: SEP,
+};
+
+/**
+ * Generates redis key for given object type and name.
+ *
+ * @param {String} type - The object type (aspect, sample, subject).
+ * @param {String} name - The object's name or absolutePath.
+ * @returns {String} the generated redis key
+ */
+function toKey(type, name) {
+  return PFX + SEP + type + SEP + name.toLowerCase();
+} // toKey
+
+/**
+ * Clear all the "sampleStore" keys (for subjects, aspects, samples) from
+ * redis.
+ *
+ * @returns {Promise} upon completion.
+ */
+function eradicate() {
+  const promises = Object.getOwnPropertyNames(constants.indexKey)
+    .map((s) => redisClient.smembersAsync(constants.indexKey[s])
+    .then((keys) => {
+      keys.push(constants.indexKey[s]);
+      return redisClient.delAsync(keys);
+    })
+    .catch((err) => {
+      // NO-OP
+      console.error(err); // eslint-disable-line no-console
+      Promise.resolve(true);
+    }));
+  return Promise.all(promises);
+} // eradicate
+
+/**
+ * Remove nulls and stringify arrays.
+ *
+ * @param {Object} obj - The object to clean.
+ * @param {Array} arrayFields - List of array fields to stringify.
+ * @returns {Object} the object with no nulls and stringified arrays.
+ */
+function removeNullsAndStringifyArrays(obj, arrayFields) {
+  Object.keys(obj).forEach((key) => {
+    if (arrayFields.includes(key)) {
+      obj[key] = JSON.stringify(obj[key]);
+    } else if (obj[key] === null) {
+      delete obj[key];
+    }
+  });
```

Comment:
If we dont want to keep null/empty arrays, then this should be changed to following:
```
 Object.keys(obj).forEach((key) => {
    // delete null fields and empty arrays, then stringify array fields
    if (obj[key] === null ||
     (Array.isArray(obj[key]) && obj[key].length === 0)) {
      delete obj[key];
    } else if (arrayFields.includes(key)) {
      obj[key] = JSON.stringify(obj[key]);
    }
  });
```